File: src/test/resources/org/bau/benchmarks/munchausen.c
// https://github.com/jabbalaci/SpeedTests/blob/master/c/main.c

#include <stdio.h>
#include <math.h>

#define MAX 440000000

int cache[10];
int is_munchausen(int number);
void set_cache();

int main() {
    set_cache();
    for (int i = 0; i < MAX; ++i)
        if (is_munchausen(i))
            printf("%d\n", i);
    return 0;
}

int is_munchausen(int number) {
    int n = number;
    int total = 0;
    while (n > 0) {
        int digit = n % 10;
        total += cache[digit];
        if (total > number)
            return 0;
        n /= 10;
    }
    return total == number;
}

void set_cache() {
    cache[0] = 0;
    for (int i = 1; i <= 9; ++i)
        cache[i] = pow(i, i);
}

File: src/test/resources/org/bau/benchmarks/munchausen.nim
# https://github.com/jabbalaci/SpeedTests

import math

func get_cache(): array[10, int] =
    result[0] = 0
    for i in 1 .. 9:
        result[i] = i ^ i

const
    MAX = 440_000_000
    cache: array[10, int] = get_cache()

func is_munchausen(number: int): bool =
    var
        n = number
        total = 0
    while n > 0:
        let digit = n mod 10
        total += cache[digit]
        if total > number:
            return false
        n = n div 10
    total == number

proc main() =
    for i in 0 ..< MAX:
        if is_munchausen(i):
            echo i

when isMainModule:
    main()